Cumulative features

pprados opened the pull request.

pprados commented:
This pull request add many features and fix some bugs:

- Add the CSV parser and dumper
- Add the comparison between grid (grid_v1 + (grid_v2 - grid_v1) == grid_v2 is True at all time)
- The README is updated
- Two new functions suffix_to_mode() and mode_to_suffix() to manage the file suffix
- Optimize the memory footprint (remove all unnecessary tag in entities in parsers)
- Fix a bug in filter, when the comparison operator is use with entity without value for the corresponding tag

I think it's possible to release a new version. I'm waiting for this version to publish another haystack project (https://github.com/pprados/haystackapi)

pprados commented:
Can you take the time to analyze this pull request?
Thanks

sjlongland commented:
@pprados Many thanks for your contributions… I've filed a sub-task for reviewing this internally and we'll see if we can slot some time in at some point in the near future.

pprados commented:
Cool ! Many thanks

I will publish some little fix next week.

pprados commented:
Ok. Now, all known bug are fixed.

sjlongland commented:
What would help greatly in reviewing this would be to split up the various changes into related subsets… "Cumulative features" and the list of commits seems to suggest this is several unrelated changes all lumped in together.

pprados commented:
Glups. I will try to reorganize this features, but it's very complex now.
You can read commit after commit. It's the best approach, I think.

pprados commented:
Hello,
It is now very difficult to redraw the code into disjointed features.  They are intimately linked.
I need all these features for another project. I want to implement the haystack REST API.
I understand that it is difficult to validate so much functionality all at once. To help you, I will describe them more precisely.

- Haystack format are three collection tag kinds: `list`, `dict` and `grid`. I've added the necessary for that
- Haystack accepts the `XStr` binary format. I’ve added this feature
- The haystack REST API must manage the CSV format. I’ve added this feature. It’s possible to use the REST API with Excel, to import the ontology and/or the time series.
- When I want to import a file into the database, I need to calculate the difference between two versions. In Grid, I’ve add the operator `__sub__()`, `__add__()` and `__eq__()`. Now, it’s possible to write
`assert v1+(v2-v1) == v2`
- Because I manipulate different files extensions, I add some methods to convert the extension to haystack type and vice versa `suffix_to_mode()` and `mode_to_suffix()`.
- With the REST API, I read large files, with many data. I want to use these data inside an AWS Lambda. It is important to save the memory used. When files are read, unnecessary tags are no longer present in memory.
- To implement the `limit` parameter in REST API, `grid[1:3]` return another grid.
- To facilitate the creation of a grid, I added functions to adjust the list of columns (`pack_columns()` and `extend_columns()`)
- To facilitate the use of entities with identifier, I have added several functionalities:
  - To be conformed to the specification, the `id` tag must be a `Ref`.
  - Grids manage an index internally
  - `grid[“myid”]`
  - `“myid” in grid`
  - `grid.keys()`
  - `del grid[“myid”]`
- I update the source to be conformed with PEP8
- I use different tools to audit the source code, and I have implemented the recommendations and bug fix
- To accept call sequencing, the methods return `self` when it’s possible

I can create a real fork, with a new project name, in order to be able to publish my HaystackAPI component, but I'd rather not do that.
It’s not a good idea to create a dependency with : `git+https://github.com/pprados/hszinc.git@diff#egg=hszinc[unitconversion]`

Tell me what you think about it.
